File: home-modules/desktop/i3-project-event-daemon/models/discovery.py
"""
Discovery models for Feature 097: Git-Based Project Discovery and Management.

Provides Pydantic models for:
- Git metadata extraction
- Repository and worktree discovery
- Scan configuration
- Discovery results

Per Constitution Principle X: Python 3.11+, Pydantic for validation.
"""

from datetime import datetime
from enum import Enum
from pathlib import Path
from typing import Optional, List
from pydantic import BaseModel, Field, field_validator


class SourceType(str, Enum):
    """Classification of how a project was discovered.

    Feature 097: All projects must be discovered - no manual creation.
    """
    LOCAL = "local"      # Standard git repository discovered on filesystem
    WORKTREE = "worktree"  # Git worktree linked to parent repository
    REMOTE = "remote"    # GitHub repository not cloned locally (listing only)


class ProjectStatus(str, Enum):
    """Current availability status of a project."""
    ACTIVE = "active"    # Directory exists and is accessible
    MISSING = "missing"  # Directory no longer exists or inaccessible


class GitMetadata(BaseModel):
    """Git-specific metadata attached to discovered projects."""

    current_branch: str = Field(
        ...,
        description="Branch name or 'HEAD' if detached"
    )
    commit_hash: str = Field(
        ...,
        min_length=7,
        max_length=7,
        description="Short SHA (7 characters)"
    )
    is_clean: bool = Field(
        ...,
        description="No uncommitted changes"
    )
    has_untracked: bool = Field(
        ...,
        description="Untracked files present"
    )
    ahead_count: int = Field(
        default=0,
        ge=0,
        description="Commits ahead of upstream"
    )
    behind_count: int = Field(
        default=0,
        ge=0,
        description="Commits behind upstream"
    )
    remote_url: Optional[str] = Field(
        default=None,
        description="Origin remote URL"
    )
    primary_language: Optional[str] = Field(
        default=None,
        description="Dominant programming language"
    )
    last_commit_date: Optional[datetime] = Field(
        default=None,
        description="Most recent commit timestamp"
    )

    @field_validator('commit_hash', mode='before')
    @classmethod
    def validate_commit_hash(cls, v: str) -> str:
        """Allow empty string for repos with no commits."""
        if v == "" or v is None:
            return "0000000"  # Placeholder for empty repos
        return v[:7] if len(v) > 7 else v


class ScanConfiguration(BaseModel):
    """User-defined settings for repository discovery."""

    scan_paths: List[str] = Field(
        min_length=1,
        description="Directories to scan for repositories"
    )
    exclude_patterns: List[str] = Field(
        default_factory=lambda: ["node_modules", "vendor", ".cache"],
        description="Directory names to skip"
    )
    auto_discover_on_startup: bool = Field(
        default=False,
        description="Run discovery when daemon starts"
    )
    max_depth: int = Field(
        default=3,
        ge=1,
        le=10,
        description="Maximum recursion depth for scanning"
    )

    @field_validator('scan_paths', mode='before')
    @classmethod
    def expand_paths(cls, v: List[str]) -> List[str]:
        """Expand ~ and resolve to absolute paths."""
        return [str(Path(p).expanduser().resolve()) for p in v]


class DiscoveredRepository(BaseModel):
    """Intermediate representation of a found repository before project creation."""

    path: str = Field(
        ...,
        description="Absolute path to repository"
    )
    name: str = Field(
        ...,
        description="Derived from directory name"
    )
    is_worktree: bool = Field(
        ...,
        description="True if .git is a file"
    )
    git_metadata: GitMetadata = Field(
        ...,
        description="Extracted git data"
    )
    parent_repo_path: Optional[str] = Field(
        default=None,
        description="For worktrees, path to main repo"
    )
    inferred_icon: str = Field(
        default="📁",
        description="Emoji based on language"
    )


class DiscoveredWorktree(DiscoveredRepository):
    """Worktree-specific discovery result with guaranteed parent reference."""

    is_worktree: bool = Field(
        default=True,
        description="Always True for worktrees"
    )
    parent_repo_path: str = Field(
        ...,
        description="Path to main repository (required for worktrees)"
    )
    inferred_icon: str = Field(
        default="🌿",
        description="Worktree icon"
    )


class SkippedPath(BaseModel):
    """Path that was skipped during discovery with reason."""

    path: str = Field(
        ...,
        description="Absolute path that was skipped"
    )
    reason: str = Field(
        ...,
        description="Reason for skipping (e.g., 'no_git_directory', 'excluded_pattern')"
    )


class DiscoveryError(BaseModel):
    """Non-fatal error encountered during discovery."""

    path: Optional[str] = Field(
        default=None,
        description="Path where error occurred (if applicable)"
    )
    source: Optional[str] = Field(
        default=None,
        description="Error source (e.g., 'github', 'filesystem')"
    )
    error: str = Field(
        ...,
        description="Error code (e.g., 'git_metadata_extraction_failed')"
    )
    message: str = Field(
        ...,
        description="Human-readable error message"
    )


class DiscoveryResult(BaseModel):
    """Ephemeral result returned from a discovery operation."""

    success: bool = Field(
        default=True,
        description="Overall success status"
    )
    discovered_repos: List[DiscoveredRepository] = Field(
        default_factory=list,
        description="Repositories found"
    )
    discovered_worktrees: List[DiscoveredWorktree] = Field(
        default_factory=list,
        description="Worktrees found"
    )
    skipped_paths: List[SkippedPath] = Field(
        default_factory=list,
        description="Paths skipped (not git repos)"
    )
    projects_created: int = Field(
        default=0,
        description="Count of new projects"
    )
    projects_updated: int = Field(
        default=0,
        description="Count of updated projects"
    )
    projects_marked_missing: int = Field(
        default=0,
        description="Count of newly missing projects"
    )
    duration_ms: int = Field(
        default=0,
        description="Time taken in milliseconds"
    )
    errors: List[DiscoveryError] = Field(
        default_factory=list,
        description="Non-fatal errors encountered"
    )


class GitHubRepo(BaseModel):
    """GitHub repository information from gh CLI."""

    name: str = Field(
        ...,
        description="Repository name"
    )
    full_name: str = Field(
        ...,
        description="Full name (owner/repo)"
    )
    description: Optional[str] = Field(
        default=None,
        description="Repository description"
    )
    primary_language: Optional[str] = Field(
        default=None,
        description="Primary programming language"
    )
    pushed_at: Optional[datetime] = Field(
        default=None,
        description="Last push timestamp"
    )
    visibility: str = Field(
        default="public",
        description="Repository visibility (public/private)"
    )
    is_private: bool = Field(
        default=False,
        description="Whether repository is private"
    )
    is_fork: bool = Field(
        default=False,
        description="Whether repository is a fork"
    )
    is_archived: bool = Field(
        default=False,
        description="Whether repository is archived"
    )
    clone_url: str = Field(
        ...,
        description="HTTPS clone URL"
    )
    ssh_url: Optional[str] = Field(
        default=None,
        description="SSH clone URL"
    )
    has_local_clone: bool = Field(
        default=False,
        description="Whether repository exists locally"
    )
    local_project_name: Optional[str] = Field(
        default=None,
        description="Name of local project if cloned"
    )

    @classmethod
    def from_gh_json(cls, data: dict) -> "GitHubRepo":
        """Create GitHubRepo from gh CLI JSON output.

        Feature 097: Parse gh repo list output format.

        Args:
            data: Dictionary from gh CLI JSON output

        Returns:
            GitHubRepo instance
        """
        # Extract primary language from nested structure
        primary_language = None
        if data.get("primaryLanguage"):
            primary_language = data["primaryLanguage"].get("name")

        # Parse pushed_at timestamp
        pushed_at = None
        if data.get("pushedAt"):
            try:
                pushed_at = datetime.fromisoformat(data["pushedAt"].replace("Z", "+00:00"))
            except (ValueError, TypeError):
                pass

        return cls(
            name=data["name"],
            full_name=data["nameWithOwner"],
            description=data.get("description"),
            primary_language=primary_language,
            pushed_at=pushed_at,
            visibility="private" if data.get("isPrivate") else "public",
            is_private=data.get("isPrivate", False),
            is_fork=data.get("isFork", False),
            is_archived=data.get("isArchived", False),
            clone_url=data.get("url", ""),
            ssh_url=data.get("sshUrl"),
        )


class GitHubListResult(BaseModel):
    """Result of listing GitHub repositories."""

    success: bool = Field(
        default=True,
        description="Whether the listing succeeded"
    )
    repos: List[GitHubRepo] = Field(
        default_factory=list,
        description="GitHub repositories"
    )
    total_count: int = Field(
        default=0,
        description="Total number of repositories"
    )
    errors: List[DiscoveryError] = Field(
        default_factory=list,
        description="Errors encountered"
    )


class CorrelationResult(BaseModel):
    """Result of correlating local and remote repositories.

    Feature 097: Identifies which GitHub repos are cloned locally
    and which are remote-only.
    """

    cloned: List[GitHubRepo] = Field(
        default_factory=list,
        description="GitHub repos that have local clones (has_local_clone=True)"
    )
    remote_only: List[GitHubRepo] = Field(
        default_factory=list,
        description="GitHub repos without local clones"
    )

File: home-modules/tools/i3pm-deno/src/models/discovery.ts
/**
 * Discovery Models for Feature 097: Git-Centric Project and Worktree Management
 *
 * TypeScript/Zod schemas for:
 * - Git metadata extraction
 * - Repository and worktree discovery
 * - Scan configuration
 * - Discovery results
 * - Panel display models
 *
 * Per Constitution Principle XIII: Deno CLI Standards with Zod 3.22+
 */

import { z } from "https://deno.land/x/zod@v3.22.4/mod.ts";

// ============================================================================
// Enums (Feature 097 - Git-centric architecture)
// ============================================================================

/**
 * Classification of project type (Feature 097).
 *
 * - repository: Primary entry point for a bare repo (ONE per bare_repo_path)
 * - worktree: Git worktree linked to a Repository Project
 * - standalone: Non-git directory OR simple repo with no worktrees
 */
export const SourceTypeSchema = z.enum(["repository", "worktree", "standalone"]);
export type SourceType = z.infer<typeof SourceTypeSchema>;

/**
 * Current availability status (Feature 097).
 *
 * - active: Directory exists and is accessible
 * - missing: Directory no longer exists or inaccessible
 * - orphaned: Worktree with no matching Repository Project
 */
export const ProjectStatusSchema = z.enum(["active", "missing", "orphaned"]);
export type ProjectStatus = z.infer<typeof ProjectStatusSchema>;

// ============================================================================
// Git Metadata
// ============================================================================

/**
 * Git-specific metadata attached to projects (Feature 097).
 */
export const GitMetadataSchema = z.object({
  /** Branch name or "HEAD" if detached */
  current_branch: z.string(),

  /** Short SHA (7 characters) */
  commit_hash: z.string().length(7),

  /** No uncommitted changes */
  is_clean: z.boolean(),

  /** Untracked files present */
  has_untracked: z.boolean(),

  /** Commits ahead of upstream */
  ahead_count: z.number().int().nonnegative().default(0),

  /** Commits behind upstream */
  behind_count: z.number().int().nonnegative().default(0),

  /** Origin remote URL */
  remote_url: z.string().nullable().default(null),

  /** Most recent file modification timestamp */
  last_modified: z.string().datetime().nullable().default(null),

  /** When metadata was last refreshed */
  last_refreshed: z.string().datetime().nullable().default(null),
});

export type GitMetadata = z.infer<typeof GitMetadataSchema>;

// ============================================================================
// Scan Configuration
// ============================================================================

/**
 * User-defined settings for repository discovery.
 */
export const ScanConfigurationSchema = z.object({
  /** Directories to scan for repositories */
  scan_paths: z.array(z.string()).min(1),

  /** Directory names to skip */
  exclude_patterns: z
    .array(z.string())
    .default(["node_modules", "vendor", ".cache"]),

  /** Run discovery when daemon starts */
  auto_discover_on_startup: z.boolean().default(false),

  /** Maximum recursion depth for scanning (1-10) */
  max_depth: z.number().int().min(1).max(10).default(3),
});

export type ScanConfiguration = z.infer<typeof ScanConfigurationSchema>;

// ============================================================================
// Discovery Results
// ============================================================================

/**
 * Intermediate representation of a found repository before project creation.
 */
export const DiscoveredRepositorySchema = z.object({
  /** Absolute path to repository */
  path: z.string(),

  /** Derived from directory name */
  name: z.string(),

  /** True if .git is a file */
  is_worktree: z.boolean(),

  /** Extracted git data */
  git_metadata: GitMetadataSchema,

  /** For worktrees, path to main repo */
  parent_repo_path: z.string().nullable().default(null),

  /** Emoji based on language */
  inferred_icon: z.string().default("📁"),
});

export type DiscoveredRepository = z.infer<typeof DiscoveredRepositorySchema>;

/**
 * Worktree-specific discovery result with guaranteed parent reference.
 */
export const DiscoveredWorktreeSchema = DiscoveredRepositorySchema.extend({
  /** Always True for worktrees */
  is_worktree: z.literal(true),

  /** Path to main repository (required for worktrees) */
  parent_repo_path: z.string(),

  /** Worktree icon */
  inferred_icon: z.string().default("🌿"),
});

export type DiscoveredWorktree = z.infer<typeof DiscoveredWorktreeSchema>;

/**
 * Path that was skipped during discovery with reason.
 */
export const SkippedPathSchema = z.object({
  /** Absolute path that was skipped */
  path: z.string(),

  /** Reason for skipping */
  reason: z.string(),
});

export type SkippedPath = z.infer<typeof SkippedPathSchema>;

/**
 * Non-fatal error encountered during discovery.
 */
export const DiscoveryErrorSchema = z.object({
  /** Path where error occurred (if applicable) */
  path: z.string().optional(),

  /** Error source (e.g., "github", "filesystem") */
  source: z.string().optional(),

  /** Error code */
  error: z.string(),

  /** Human-readable error message */
  message: z.string(),
});

export type DiscoveryError = z.infer<typeof DiscoveryErrorSchema>;

/**
 * Ephemeral result returned from a discovery operation.
 */
export const DiscoveryResultSchema = z.object({
  /** Overall success status */
  success: z.boolean().default(true),

  /** Repositories found */
  discovered_repos: z.array(DiscoveredRepositorySchema).default([]),

  /** Worktrees found */
  discovered_worktrees: z.array(DiscoveredWorktreeSchema).default([]),

  /** Paths skipped (not git repos) */
  skipped_paths: z.array(SkippedPathSchema).default([]),

  /** Count of new projects */
  projects_created: z.number().int().default(0),

  /** Count of updated projects */
  projects_updated: z.number().int().default(0),

  /** Count of newly missing projects */
  projects_marked_missing: z.number().int().default(0),

  /** Time taken in milliseconds */
  duration_ms: z.number().int().default(0),

  /** Non-fatal errors encountered */
  errors: z.array(DiscoveryErrorSchema).default([]),
});

export type DiscoveryResult = z.infer<typeof DiscoveryResultSchema>;

// ============================================================================
// GitHub Integration
// ============================================================================

/**
 * GitHub repository information from gh CLI.
 */
export const GitHubRepoSchema = z.object({
  /** Repository name */
  name: z.string(),

  /** Full name (owner/repo) */
  full_name: z.string(),

  /** Repository description */
  description: z.string().nullable().default(null),

  /** Primary programming language */
  primary_language: z.string().nullable().default(null),

  /** Last push timestamp */
  pushed_at: z.string().datetime().nullable().default(null),

  /** Repository visibility */
  visibility: z.enum(["public", "private"]).default("public"),

  /** Whether repository is a fork */
  is_fork: z.boolean().default(false),

  /** Whether repository is archived */
  is_archived: z.boolean().default(false),

  /** HTTPS clone URL */
  clone_url: z.string(),

  /** Whether repository exists locally */
  has_local_clone: z.boolean().default(false),

  /** Name of local project if cloned */
  local_project_name: z.string().nullable().default(null),
});

export type GitHubRepo = z.infer<typeof GitHubRepoSchema>;

/**
 * Result of listing GitHub repositories.
 */
export const GitHubListResultSchema = z.object({
  /** Whether the listing succeeded */
  success: z.boolean().default(true),

  /** GitHub repositories */
  repos: z.array(GitHubRepoSchema).default([]),

  /** Total number of repositories */
  total_count: z.number().int().default(0),

  /** Errors encountered */
  errors: z.array(DiscoveryErrorSchema).default([]),
});

export type GitHubListResult = z.infer<typeof GitHubListResultSchema>;

// ============================================================================
// Feature 097: Unified Project Schema
// ============================================================================

/**
 * Remote SSH configuration for projects.
 */
export const RemoteConfigSchema = z.object({
  enabled: z.boolean().default(false),
  host: z.string().default(""),
  user: z.string().default(""),
  remote_dir: z.string().default(""),
  port: z.number().int().min(1).max(65535).default(22),
});

export type RemoteConfig = z.infer<typeof RemoteConfigSchema>;

/**
 * Unified Project model for Feature 097 git-centric architecture.
 *
 * This is the single schema for all project types: repository, worktree, standalone.
 */
export const ProjectSchema = z.object({
  /** Unique project identifier (slug) */
  name: z.string().min(1).max(64),

  /** Human-readable display name */
  display_name: z.string().min(1),

  /** Absolute path to project directory */
  directory: z.string().min(1),

  /** Emoji icon for visual identification */
  icon: z.string().default("📁"),

  /** Window hiding behavior */
  scope: z.enum(["scoped", "global"]).default("scoped"),

  /** Remote SSH configuration */
  remote: RemoteConfigSchema.nullable().default(null),

  // Feature 097: Git-centric fields
  /** Project type: repository (primary), worktree (linked), or standalone (non-git) */
  source_type: SourceTypeSchema.default("standalone"),

  /** Availability status: active, missing, or orphaned */
  status: ProjectStatusSchema.default("active"),

  /** GIT_COMMON_DIR - canonical identifier for all worktrees of a repo */
  bare_repo_path: z.string().nullable().default(null),

  /** For worktrees: name of the parent Repository Project */
  parent_project: z.string().nullable().default(null),

  /** Cached git state */
  git_metadata: GitMetadataSchema.nullable().default(null),

  /** App window classes scoped to this project */
  scoped_classes: z.array(z.string()).default([]),

  /** When project was created */
  created_at: z.string().datetime().nullable().default(null),

  /** When project was last modified */
  updated_at: z.string().datetime().nullable().default(null),
}).refine(
  (data) => {
    // Feature 097 T005 equivalent: Worktree projects must have parent_project
    if (data.source_type === "worktree" && !data.parent_project) {
      return false;
    }
    return true;
  },
  { message: "Worktree projects must have parent_project set" }
);

export type Project = z.infer<typeof ProjectSchema>;

// ============================================================================
// Feature 097: Panel Display Models
// ============================================================================

/**
 * Repository project with its child worktrees for panel display (T014 equivalent).
 */
export const RepositoryWithWorktreesSchema = z.object({
  /** The repository project (source_type=repository) */
  project: ProjectSchema,

  /** Number of child worktrees */
  worktree_count: z.number().int().nonnegative().default(0),

  /** True if any worktree has uncommitted changes */
  has_dirty: z.boolean().default(false),

  /** UI expansion state */
  is_expanded: z.boolean().default(true),

  /** Child worktree projects */
  worktrees: z.array(ProjectSchema).default([]),
});

export type RepositoryWithWorktrees = z.infer<typeof RepositoryWithWorktreesSchema>;

/**
 * Complete panel data structure (T013 equivalent).
 */
export const PanelProjectsDataSchema = z.object({
  /** Repository projects with their grouped worktrees */
  repository_projects: z.array(RepositoryWithWorktreesSchema).default([]),

  /** Standalone projects (non-git or simple repos) */
  standalone_projects: z.array(ProjectSchema).default([]),

  /** Worktrees with no matching Repository Project */
  orphaned_worktrees: z.array(ProjectSchema).default([]),

  /** Currently active project name */
  active_project: z.string().nullable().default(null),
});

export type PanelProjectsData = z.infer<typeof PanelProjectsDataSchema>;

// ============================================================================
// RPC Request/Response Types
// ============================================================================

/**
 * Parameters for discover_projects RPC method.
 */
export const DiscoverProjectsParamsSchema = z.object({
  /** Override scan paths from config */
  paths: z.array(z.string()).optional(),

  /** Also query GitHub repos */
  include_github: z.boolean().default(false),

  /** Report without creating projects */
  dry_run: z.boolean().default(false),
});

export type DiscoverProjectsParams = z.infer<typeof DiscoverProjectsParamsSchema>;

/**
 * Parameters for get_discovery_config RPC method.
 */
export const GetDiscoveryConfigParamsSchema = z.object({}).optional();

export type GetDiscoveryConfigParams = z.infer<typeof GetDiscoveryConfigParamsSchema>;

/**
 * Parameters for update_discovery_config RPC method.
 */
export const UpdateDiscoveryConfigParamsSchema = z.object({
  /** Directories to scan */
  scan_paths: z.array(z.string()).optional(),

  /** Directory names to skip */
  exclude_patterns: z.array(z.string()).optional(),

  /** Enable startup discovery */
  auto_discover_on_startup: z.boolean().optional(),

  /** Max recursion depth (1-10) */
  max_depth: z.number().int().min(1).max(10).optional(),
});

export type UpdateDiscoveryConfigParams = z.infer<typeof UpdateDiscoveryConfigParamsSchema>;

/**
 * Parameters for refresh_git_metadata RPC method.
 */
export const RefreshGitMetadataParamsSchema = z.object({
  /** Specific project to refresh (all if omitted) */
  project_name: z.string().optional(),
});

export type RefreshGitMetadataParams = z.infer<typeof RefreshGitMetadataParamsSchema>;

/**
 * Parameters for list_github_repos RPC method.
 */
export const ListGitHubReposParamsSchema = z.object({
  /** Maximum repos to return */
  limit: z.number().int().default(100),

  /** Include archived repos */
  include_archived: z.boolean().default(false),

  /** Include forked repos */
  include_forks: z.boolean().default(false),
});

export type ListGitHubReposParams = z.infer<typeof ListGitHubReposParamsSchema>;
